compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp

```diff
@@ -0,0 +1,411 @@
+//
+// These allow us to use a default channel for all communications, and there is
+// 1:1 mapping from the replica IDs to the communication ranks. The attribute,
+// use_global_device_ids, is always set in this case.
+//
+// The next stage is to support multiple replica groups. This needs a channel
+// creation with a subset of processes, which should have another communication
+// among the group. A possible strategy is to have the root process in the group
+// (the first rank of the group) creates a channel and the other processes query
+// the channel info from the root process. A key-value store using gRPC might be
+// a good solution.
+//
+// Supporting partition_id comes next. This includes the support for various
+// mode combinations for cross-replica and cross partition communication. See
+// the stablehlo specification for more details about the different modes.
+
+namespace {
+
+static std::optional<IREE::Flow::CollectiveElementType>
+convertToFlowCollectiveElementType(Type type) {
+  if (type.isF32()) return IREE::Flow::CollectiveElementType::Float32;
+
+  if (type.isInteger(32)) {
+    if (type.isSignedInteger())
+      return IREE::Flow::CollectiveElementType::Sint32;
+    else
+      return IREE::Flow::CollectiveElementType::Uint32;
+  }
+
+  if (type.isF16()) return IREE::Flow::CollectiveElementType::Float16;
+
+  if (type.isInteger(8)) {
+    if (type.isSignedInteger())
+      return IREE::Flow::CollectiveElementType::Sint8;
+    else
+      return IREE::Flow::CollectiveElementType::Uint8;
+  }
+
+  if (type.isInteger(16)) {
+    if (type.isSignedInteger())
+      return IREE::Flow::CollectiveElementType::Sint16;
+    else
+      return IREE::Flow::CollectiveElementType::Uint16;
+  }
+
+  if (type.isBF16()) return IREE::Flow::CollectiveElementType::BFloat16;
+
+  if (type.isF64()) return IREE::Flow::CollectiveElementType::Float64;
+
+  if (type.isInteger(64)) {
+    if (type.isSignedInteger())
+      return IREE::Flow::CollectiveElementType::Sint64;
+    else
+      return IREE::Flow::CollectiveElementType::Uint64;
+  }
+
+  return std::nullopt;
+}
+
+static std::optional<IREE::Flow::CollectiveReductionOp>
+convertToFlowCollectiveReductionOp(const Operation &op) {
+  if (isa<mhlo::AddOp>(op)) {
+    return IREE::Flow::CollectiveReductionOp::ReductionSum;
+  } else if (isa<mhlo::MulOp>(op)) {
+    return IREE::Flow::CollectiveReductionOp::ReductionProduct;
+  } else if (isa<mhlo::MinOp>(op)) {
+    return IREE::Flow::CollectiveReductionOp::ReductionMinimum;
+  } else if (isa<mhlo::MaxOp>(op)) {
+    return IREE::Flow::CollectiveReductionOp::ReductionMaximum;
+  } else {
+    // TODO: we may be able to detect an average operation and convert it
+    // into IREE::Flow::CollectiveReductionOp::ReductionAverage.
+    return std::nullopt;
+  }
+}
+
+}  // namespace
+
+/// Converts mhlo.replica_id to flow.channel.default + flow.channel.rank.
+/// TODO: this assumes that there is no partition so that there is a 1:1 mapping
```

Comment:
use either issues or area owner for todos, like `// TODO(okkwon):` so that people in the future (or yourself) can know who holds context on the work to be done.

https://google.github.io/styleguide/cppguide.html#TODO_Comments